File: 백준/Gold/2096. 내려가기/내려가기.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
)

var (
	reader = bufio.NewReader(os.Stdin)
	writer = bufio.NewWriter(os.Stdout)
)

func main() {
	defer writer.Flush()
	var n int
	var a, b, c float64
	var min, max float64
	fmt.Fscanf(reader, "%d\n", &n)
	var maxRes, minRes, nowMax, nowMin = make([]float64, 3), make([]float64, 3), make([]float64, 3), make([]float64, 3)

	for i := 0; i < n; i++ {
		fmt.Fscanf(reader, "%f %f %f\n", &a, &b, &c)
		for j := 0; j < 3; j++ {
			if j == 0 {
				nowMax[0] = a + math.Max(maxRes[0], maxRes[1])
				nowMin[0] = a + math.Min(minRes[0], minRes[1])
			} else if j == 1 {
				nowMax[1] = b + math.Max(maxRes[0], math.Max(maxRes[1], maxRes[2]))
				nowMin[1] = b + math.Min(minRes[0], math.Min(minRes[1], minRes[2]))
			} else if j == 2 {
				nowMax[2] = c + math.Max(maxRes[1], maxRes[2])
				nowMin[2] = c + math.Min(minRes[1], minRes[2])
			}
		}
		for j := 0; j < 3; j++ {
			maxRes[j] = nowMax[j]
			minRes[j] = nowMin[j]
		}
	}
	max = math.Max(maxRes[0], math.Max(maxRes[1], maxRes[2]))
	min = math.Min(minRes[0], math.Min(minRes[1], minRes[2]))
	fmt.Fprintf(writer, "%0.f %0.f\n", max, min)
}

File: 백준/Gold/2096. 내려가기/내려가기.py
import sys 
input= sys.stdin.readline
sys.setrecursionlimit(10**6)

n = int(input())

max_res = [0 for _ in range(3)]
min_res = [0 for _ in range(3)]
now_max = [0 for _ in range(3)]
now_min = [0 for _ in range(3)]
for _ in range(n):
    tmp = list(map(int, input().split()))
    for i in range(3):
        if i == 0:
            now_max[0] = tmp[0] + max(max_res[0], max_res[1])
            now_min[0] = tmp[0] + min(min_res[0], min_res[1])
        elif i == 1:
            now_max[1] = tmp[1] + max(max_res[0], max_res[1], max_res[2])
            now_min[1] = tmp[1] + min(min_res[0], min_res[1], min_res[2])
        else:
            now_max[2] = tmp[2]+ max(max_res[1], max_res[2])
            now_min[2] = tmp[2]+ min(min_res[1], min_res[2])

    for i in range(3):
        max_res[i] = now_max[i]
        min_res[i] = now_min[i]

print(max(max_res), min(min_res))